src/tactic/compute_degree.lean

```diff
@@ -0,0 +1,176 @@
+* Add better functionality to deal with exponents that are not necessarily closed natural numbers.
+* Add support for proving goals of the from `f.(nat_)degree ≠ 0`.
+* Make sure that `degree` and `nat_degree` are equally supported.
+
+##  Implementation details
+
+We start with a goal of the form `f.(nat_)degree ≤ d`.  Recurse into `f` breaking apart sums,
+products and powers.  Take care of numerals, `C a, X (^ n), monomial a n` separately. -/
+
+namespace polynomial
+variables {R : Type*}
+
+lemma nat_degree_sub_le_iff_left [ring R] {n : ℕ} (p q : polynomial R) (qn : q.nat_degree ≤ n) :
+  (p - q).nat_degree ≤ n ↔ p.nat_degree ≤ n :=
+begin
+  rw [sub_eq_add_neg, nat_degree_add_le_iff_left],
+  rwa nat_degree_neg,
+end
+
+variables [semiring R] (a : polynomial R)
+
+lemma nat_degree_bit0 : (bit0 a).nat_degree ≤ a.nat_degree :=
+(nat_degree_add_le _ _).trans (max_self _).le
+
+lemma nat_degree_bit1 : (bit1 a).nat_degree ≤ a.nat_degree :=
+(nat_degree_add_le _ _).trans (by simp [nat_degree_bit0])
+
+end polynomial
+
+namespace tactic
+namespace compute_degree
+open expr polynomial
+
+/--  `guess_degree e` assumes that `e` is an expression in a polynomial ring, and makes an attempt
+at guessing the degree of `e`.  Heuristics for `guess_degree`:
+* `0, 1, C a`,      guess `0`,
+* `polynomial.X`,   guess `1`,
+*  `bit0/1 f, -f`,  guess `guess_degree f`,
+* `f + g, f - g`,   guess `max (guess_degree f) (guess_degree g)`,
+* `f * g`,          guess `guess_degree f + guess_degree g`,
+* `f ^ n`,          guess `guess_degree f * n`,
+* `monomial n r`,   guess `n`,
+* `f` not as above, guess `f.nat_degree`.
+ -/
+meta def guess_degree : expr → tactic expr
+| `(has_zero.zero)           := pure `(0)
+| `(has_one.one)             := pure `(0)
+| `(- %%f)                   := guess_degree f
+| (app `(⇑C) x)              := pure `(0)
+| `(X)                       := pure `(1)
+| `(bit0 %%a)                := guess_degree a
+| `(bit1 %%a)                := guess_degree a
+| `(%%a + %%b)               := do [da, db] ← [a, b].mmap guess_degree,
+                                pure $ expr.mk_app `(max : ℕ → ℕ → ℕ) [da, db]
+| `(%%a - %%b)               := do [da, db] ← [a, b].mmap guess_degree,
+                                pure $ expr.mk_app `(max : ℕ → ℕ → ℕ) [da, db]
+| `(%%a * %%b)               := do [da, db] ← [a, b].mmap guess_degree,
+                                pure $ expr.mk_app `((+) : ℕ → ℕ → ℕ) [da, db]
+| `(%%a ^ %%b)               := do da ← guess_degree a,
+                                pure $ expr.mk_app `((*) : ℕ → ℕ → ℕ) [da, b]
+| (app `(⇑(monomial %%n)) x) := pure n
+| e                          := do `(@polynomial %%R %%inst) ← infer_type e,
+                                pe ← to_expr ``(@nat_degree %%R %%inst) tt ff,
+                                pure $ expr.mk_app pe [e]
+
+/--  `resolve_sum_step tf e` takes a boolean `tf` and an expression `e` as inputs.
```

Comment:
```suggestion
/--  `resolve_sum_step tf e` takes an expression `e` as input.
```
? I don't see `tf`

Reply:
Thanks for catching this mistake!

I had an earlier implementation where the boolean distinguished two tactic behaviours when there were powers.  I merged the two branches, but forgot to update the doc-string.